pkg/controller/servicemesh/webhooks/mutation/controlplane.go

```diff
@@ -35,52 +39,50 @@ var _ admission.DecoderInjector = (*ControlPlaneMutator)(nil)
 func (v *ControlPlaneMutator) Handle(ctx context.Context, req admission.Request) admission.Response {
 	log := logf.Log.WithName("smcp-mutator").
 		WithValues("ServiceMeshControlPlane", webhookcommon.ToNamespacedName(&req.AdmissionRequest))
-	smcp := &maistrav1.ServiceMeshControlPlane{}
 
-	err := v.decoder.Decode(req, smcp)
+	// do we care about this object?
+	if !v.namespaceFilter.Watching(req.Namespace) {
+		log.Info(fmt.Sprintf("operator is not watching namespace '%s'", req.Namespace))
+		return admission.Allowed("")
+	}
+
+	mutator, err := v.decodeRequest(req, log)
 	if err != nil {
 		log.Error(err, "error decoding admission request")
 		return admission.Errored(http.StatusBadRequest, err)
-	} else if smcp.ObjectMeta.DeletionTimestamp != nil {
+	} else if mutator.Object().GetDeletionTimestamp() != nil {
 		log.Info("skipping deleted smcp resource")
 		return admission.Allowed("")
 	}
 
-	// do we care about this object?
-	if !v.namespaceFilter.Watching(smcp.Namespace) {
-		log.Info(fmt.Sprintf("operator is not watching namespace '%s'", smcp.Namespace))
-		return admission.Allowed("")
-	}
-
-	newSmcp := smcp.DeepCopy()
-	smcpMutated := false
-
 	// on create we set the version to the current default version
 	// on update we leave the version intact to preserve the v1.0 version
 	// implied by the missing version field, which we added in version v1.1
-	if smcp.Spec.Version == "" {
+	currentVersion := mutator.NewVersion()
+	if currentVersion == "" {
 		switch req.AdmissionRequest.Operation {
 		case admissionv1beta1.Create:
 			log.Info("Setting .spec.version to default value", "version", versions.DefaultVersion.String())
-			newSmcp.Spec.Version = versions.DefaultVersion.String()
-			smcpMutated = true
+			mutator.SetVersion(versions.DefaultVersion.String())
 		case admissionv1beta1.Update:
-			log.Info("Setting .spec.version to default value", "version", smcp.Spec.Version)
-			newSmcp.Spec.Version = smcp.Spec.Version
-			smcpMutated = true
+			oldVersion := mutator.OldVersion()
+			if currentVersion != oldVersion && oldVersion != versions.InvalidVersion.String() {
+				log.Info("Setting .spec.version to existing value", "version", oldVersion)
+				mutator.SetVersion(oldVersion)
+			}
 		}
 	}
 
-	if smcp.Spec.Template == "" {
-		log.Info("Setting .spec.template to default value", "template", maistrav1.DefaultTemplate)
-		newSmcp.Spec.Template = maistrav1.DefaultTemplate
-		smcpMutated = true
+	if len(mutator.GetProfiles()) == 0 {
+		log.Info("Setting .spec.profiles to default value", "profiles")
```

Comment:
Missing argument. The function call panics with the following message: 
```
odd number of arguments passed as key-value pairs for logging   {"ServiceMeshControlPlane": "istio-system/minimal-install", "ignored key": "profiles"}
```

Reply:
fixed